semantics/src/main/scala/aqua/semantics/rules/types/TypesInterpreter.scala

```diff
@@ -415,6 +416,44 @@ class TypesInterpreter[S[_], X](using
       }
     }
 
+  override def ensureTypeConstructibleFrom(
+    token: Token[S],
+    expected: AbilityType | StructType,
+    arguments: NonEmptyMapImpl.Type[String, (NamedArg[S], Type)]
+  ): State[X, Boolean] = for {
+    enough <- expected.fields.toNel.traverse { case (name, typ) =>
+      arguments.lookup(name) match {
+        case Some(arg -> givenType) =>
+          ensureTypeMatches(arg.argValue, typ, givenType)
+        case None =>
+          report
+            .error(
+              token,
+              s"Missing argument '$name' of type '$typ'"
+            )
+            .as(false)
+      }
+    }.map(_.forall(identity))
+    expectedKeys = expected.fields.keys.toNonEmptyList
+    _ <- arguments.toNel.traverse_ { case (name, arg -> typ) =>
+      expected.fields.lookup(name) match {
```

Comment:
can you add some comments of what is happening in this function. Maybe some small example will be great too